File: src/quicksort.py
'''
POC for quicksort
'''


def partition(A: [], p: int, r:int) -> int:
    pivot = A[r]
    i = p - 1
    for j in range(p, r):
        if A[j] <= pivot:
            i += 1
            A[i], A[j] = A[j], A[i]
    i += 1
    A[i], A[r] = A[r], A[i]
    return i

def quicksort(A: [], p:int, r: int):
    if p < r:
        q = partition(A, p, r)
        quicksort(A, p, q-1)
        quicksort(A, q+1, r)

arr = [0, 10, 7, 5, 3, 2, 200, 50, 17, 32, 9]
n = len(arr)-1
quicksort(arr, 0, n)

print(arr)

File: src/quicksort.rb

def partition(array, p, r)
  pivot = array[r]
  i = p-1

  j = p
  until j == r do
    if array[j] <= pivot
      i += 1
      array[i], array[j] = array[j], array[i]
    end
    j += 1
  end

  i += 1
  array[i], array[r] = array[r], array[i]

  i
end

def quicksort(array, p, r)
  if p < r
    q = partition(array, p, r)
    quicksort(array, p, q-1)
    quicksort(array, q+1, r)
  end
  array
end

arr = [1, 7, 3, 0, 10, 50, 30, 2]

n = arr.size - 1
puts(quicksort(arr, 0, n))
